File: server/etl_vitalcare.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

// processFinancialDataVitalCare reads a CSV file and processes financial data for VitalCare.
func processFinancialDataVitalCare(filePath string, amountIdx int) (map[string]map[string]int64, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("error opening financial analysis file: %v", err)
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.FieldsPerRecord = -1

	if _, err := r.Read(); err != nil {
		return nil, fmt.Errorf("error reading CSV header: %v", err)
	}

	data := make(map[string]map[string]int64)

	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("error reading CSV record: %v", err)
		}

		providerName := strings.TrimSpace(record[0])
		dateStr := record[1]
		amountStr := record[amountIdx]

		parts := strings.Split(dateStr, "_")
		if len(parts) != 2 {
			continue
		}
		month := strings.ToLower(parts[0])

		floatAmount, err := strconv.ParseFloat(amountStr, 64)
		if err != nil {
			log.Printf("Could not parse amount: %s", amountStr)
			continue
		}
		amount := int64(floatAmount)

		if _, ok := data[month]; !ok {
			data[month] = make(map[string]int64)
		}

		data[month][providerName] += amount
	}

	return data, nil
}

// processFinancialAnalysisVitalCare processes financial analysis data for VitalCare.
func processFinancialAnalysisVitalCare(filePath string, amountIdx int) (map[string]map[string]map[string]float64, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("error opening financial analysis file: %v", err)
	}
	defer f.Close()

	r := csv.NewReader(f)
	if _, err := r.Read(); err != nil {
		return nil, fmt.Errorf("error reading CSV header: %v", err)
	}

	data := make(map[string]map[string]map[string]float64)

	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue
		}

		providerName := strings.TrimSpace(record[0])
		monthStr := strings.TrimSpace(record[1])
		cptCode := strings.TrimSpace(record[2])
		amountStr := record[amountIdx]

		parts := strings.Split(monthStr, "_")
		if len(parts) != 2 {
			continue
		}
		month := strings.ToLower(parts[0])

		amount, err := strconv.ParseFloat(strings.TrimSpace(amountStr), 64)
		if err != nil {
			continue
		}

		if _, ok := data[month]; !ok {
			data[month] = make(map[string]map[string]float64)
		}
		if _, ok := data[month][cptCode]; !ok {
			data[month][cptCode] = make(map[string]float64)
		}

		data[month][cptCode][providerName] += amount
	}

	return data, nil
}

// processRVUsVitalCare reads RVU data for VitalCare.
func processRVUsVitalCare(filePath string) (map[string]map[string]int64, error) {
	return processFinancialDataVitalCare(filePath, 7)
}

// processTotalVisitsVitalCare reads total visits data for VitalCare.
func processTotalVisitsVitalCare(filePath string) (map[string]map[string]int64, error) {
	return processFinancialDataVitalCare(filePath, 6)
}

// processChargesVitalCare processes charges data for VitalCare.
func processChargesVitalCare(filePath string) (map[string]map[string]map[string]float64, error) {
	return processFinancialAnalysisVitalCare(filePath, 4)
}

// processPaymentsVitalCare processes payments data for VitalCare.
func processPaymentsVitalCare(filePath string) (map[string]map[string]map[string]float64, error) {
	return processFinancialAnalysisVitalCare(filePath, 7)
}

// processContractualAdjustmentsVitalCare processes contractual adjustments for VitalCare.
func processContractualAdjustmentsVitalCare(filePath string) (map[string]map[string]map[string]float64, error) {
	return processFinancialAnalysisVitalCare(filePath, 11)
}

// processUnitsVitalCare processes units data for VitalCare.
func processUnitsVitalCare(filePath string) (map[string]map[string]map[string]float64, error) {
	return processFinancialAnalysisVitalCare(filePath, 17)
}

// processPayrollVitalCare processes payroll data for VitalCare.
func processPayrollVitalCare(filePath string) (map[string]map[string]float64, []string, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, nil, fmt.Errorf("error opening payroll file: %v", err)
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.FieldsPerRecord = -1

	for i := 0; i < 4; i++ {
		if _, err := r.Read(); err != nil {
			return nil, nil, fmt.Errorf("error reading CSV header: %v", err)
		}
	}

	data := make(map[string]map[string]float64)
	var currentEmployee string
	employeeSet := make(map[string]struct{})

	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("Error reading CSV record: %v", err)
			continue
		}

		if len(record) < 15 {
			continue
		}

		if strings.TrimSpace(record[4]) != "" {
			currentEmployee = strings.Trim(strings.TrimSpace(record[4]), "\"")
			employeeSet[currentEmployee] = struct{}{}
		}

		if currentEmployee == "" {
			continue
		}

		dateStr := strings.TrimSpace(record[10])
		netPayStr := strings.TrimSpace(record[14])

		if dateStr == "" || netPayStr == "" {
			continue
		}

		parsedTime, err := time.Parse("01/02/2006", dateStr)
		if err != nil {
			log.Printf("Could not parse date: %s", dateStr)
			continue
		}
		month := strings.ToLower(parsedTime.Month().String())

		netPay, err := strconv.ParseFloat(netPayStr, 64)
		if err != nil {
			log.Printf("Could not parse net pay: %s", netPayStr)
			continue
		}

		if _, ok := data[month]; !ok {
			data[month] = make(map[string]float64)
		}

		data[month][currentEmployee] += netPay
	}

	var uniqueEmployees []string
	for employee := range employeeSet {
		uniqueEmployees = append(uniqueEmployees, strings.TrimSpace(employee))
	}

	return data, uniqueEmployees, nil
}

// getUniqueCPTCodesVitalCare gets unique CPT codes for VitalCare.
func getUniqueCPTCodesVitalCare(filePath string) ([]string, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("error opening financial analysis file: %v", err)
	}
	defer f.Close()

	r := csv.NewReader(f)
	if _, err := r.Read(); err != nil {
		return nil, fmt.Errorf("error reading CSV header: %v", err)
	}

	cptCodeSet := make(map[string]struct{})
	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue
		}

		cptCode := record[2]
		cptCodeSet[cptCode] = struct{}{}
	}

	uniqueCPTCodes := make([]string, 0, len(cptCodeSet))
	for code := range cptCodeSet {
		uniqueCPTCodes = append(uniqueCPTCodes, code)
	}

	return uniqueCPTCodes, nil
}

// processProviderLocationRelationshipVitalCare processes provider-location relationships for VitalCare.
func processProviderLocationRelationshipVitalCare(filePath string) (map[string][]string, []string, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, nil, fmt.Errorf("failed opening provider location relationship file: %w", err)
	}
	defer f.Close()

	r := csv.NewReader(f)

	_, err = r.Read()
	if err != nil {
		return nil, nil, fmt.Errorf("error reading CSV header: %v", err)
	}

	locationProviderRelationships := make(map[string][]string)
	providerSet := make(map[string]struct{}) // Use a map to store unique provider names

	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, nil, fmt.Errorf("failed reading provider location relationship file: %w", err)
		}

		providerName := strings.TrimSpace(record[0])
		locationName := strings.TrimSpace(record[1])

		locationProviderRelationships[locationName] = append(locationProviderRelationships[locationName], providerName)
		providerSet[providerName] = struct{}{}
	}

	var uniqueProviders []string
	for provider := range providerSet {
		uniqueProviders = append(uniqueProviders, strings.TrimSpace(provider))
	}

	return locationProviderRelationships, uniqueProviders, nil
}

File: server/src/etl/etlVitalcare.ts
import fs from "fs";
import csv from "csv-parser";
import { MONTHS } from "../types/common";

/**
 * Processes financial data from CSV for VitalCare
 */
export async function processFinancialDataVitalCare(
  filePath: string,
  amountIdx: number
): Promise<Record<string, Record<string, number>>> {
  return new Promise((resolve, reject) => {
    const data: Record<string, Record<string, number>> = {};

    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row: any) => {
        const values = Object.values(row) as string[];
        const providerName = values[0]?.trim();
        const dateStr = values[1];
        const amountStr = values[amountIdx];

        if (!dateStr) return;

        const parts = dateStr.split("_");
        if (parts.length !== 2) return;
        const month = parts[0].toLowerCase();

        const floatAmount = parseFloat(amountStr);
        if (isNaN(floatAmount)) {
          console.log(`Could not parse amount: ${amountStr}`);
          return;
        }
        const amount = Math.floor(floatAmount);

        if (!data[month]) {
          data[month] = {};
        }

        data[month][providerName] = (data[month][providerName] || 0) + amount;
      })
      .on("end", () => resolve(data))
      .on("error", (err) => reject(err));
  });
}

/**
 * Processes financial analysis data for VitalCare
 */
export async function processFinancialAnalysisVitalCare(
  filePath: string,
  amountIdx: number
): Promise<Record<string, Record<string, Record<string, number>>>> {
  return new Promise((resolve, reject) => {
    const data: Record<string, Record<string, Record<string, number>>> = {};

    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row: any) => {
        const values = Object.values(row) as string[];
        const providerName = values[0]?.trim();
        const monthStr = values[1]?.trim();
        const cptCode = values[2]?.trim();
        const amountStr = values[amountIdx];

        if (!monthStr) return;

        const parts = monthStr.split("_");
        if (parts.length !== 2) return;
        const month = parts[0].toLowerCase();

        const amount = parseFloat(amountStr?.trim());
        if (isNaN(amount)) return;

        if (!data[month]) {
          data[month] = {};
        }
        if (!data[month][cptCode]) {
          data[month][cptCode] = {};
        }

        data[month][cptCode][providerName] =
          (data[month][cptCode][providerName] || 0) + amount;
      })
      .on("end", () => resolve(data))
      .on("error", (err) => reject(err));
  });
}

/**
 * Processes RVU data for VitalCare
 */
export async function processRVUsVitalCare(
  filePath: string
): Promise<Record<string, Record<string, number>>> {
  return processFinancialDataVitalCare(filePath, 7);
}

/**
 * Processes total visits data for VitalCare
 */
export async function processTotalVisitsVitalCare(
  filePath: string
): Promise<Record<string, Record<string, number>>> {
  return processFinancialDataVitalCare(filePath, 6);
}

/**
 * Processes monthly patient count
 */
export async function processMonthlyPatientCount(
  filePath: string
): Promise<Record<string, Record<string, number>>> {
  return processFinancialDataVitalCare(filePath, 16);
}

/**
 * Processes charges data for VitalCare
 */
export async function processChargesVitalCare(
  filePath: string
): Promise<Record<string, Record<string, Record<string, number>>>> {
  return processFinancialAnalysisVitalCare(filePath, 4);
}

/**
 * Processes payments data for VitalCare
 */
export async function processPaymentsVitalCare(
  filePath: string
): Promise<Record<string, Record<string, Record<string, number>>>> {
  return processFinancialAnalysisVitalCare(filePath, 7);
}

/**
 * Processes contractual adjustments for VitalCare
 */
export async function processContractualAdjustmentsVitalCare(
  filePath: string
): Promise<Record<string, Record<string, Record<string, number>>>> {
  return processFinancialAnalysisVitalCare(filePath, 11);
}

/**
 * Processes units data for VitalCare
 */
export async function processUnitsVitalCare(
  filePath: string
): Promise<Record<string, Record<string, Record<string, number>>>> {
  return processFinancialAnalysisVitalCare(filePath, 17);
}

/**
 * Processes payer payment data for VitalCare
 */
export async function processPayerPaymentVitalCare(
  filePath: string
): Promise<Record<string, Record<string, number>>> {
  return processFinancialDataVitalCare(filePath, 9);
}

/**
 * Processes patient payment data for VitalCare
 */
export async function processPatientPaymentVitalCare(
  filePath: string
): Promise<Record<string, Record<string, number>>> {
  return processFinancialDataVitalCare(filePath, 10);
}

/**
 * Processes payroll data for VitalCare
 */
export async function processPayrollVitalCare(filePath: string): Promise<{
  data: Record<string, Record<string, number>>;
  uniqueEmployees: string[];
}> {
  return new Promise((resolve, reject) => {
    const data: Record<string, Record<string, number>> = {};
    const employeeSet = new Set<string>();
    let currentEmployee = "";
    let skipCount = 0;

    fs.createReadStream(filePath)
      .pipe(csv({ headers: false }))
      .on("data", (row: any) => {
        // Skip first 4 header rows
        if (skipCount < 4) {
          skipCount++;
          return;
        }

        const values = Object.values(row) as string[];

        if (values.length < 15) return;

        // Check if this is a new employee row
        if (values[4]?.trim()) {
          currentEmployee = values[4].trim().replace(/"/g, "");
          employeeSet.add(currentEmployee);
        }

        if (!currentEmployee) return;

        const dateStr = values[10]?.trim();
        const netPayStr = values[14]?.trim();

        if (!dateStr || !netPayStr) return;

        const parsedDate = new Date(dateStr);
        if (isNaN(parsedDate.getTime())) {
          console.log(`Could not parse date: ${dateStr}`);
          return;
        }
        const month = MONTHS[parsedDate.getMonth()];

        const netPay = parseFloat(netPayStr);
        if (isNaN(netPay)) {
          console.log(`Could not parse net pay: ${netPayStr}`);
          return;
        }

        if (!data[month]) {
          data[month] = {};
        }

        data[month][currentEmployee] =
          (data[month][currentEmployee] || 0) + netPay;
      })
      .on("end", () => {
        const uniqueEmployees = Array.from(employeeSet).map((e) => e.trim());
        resolve({ data, uniqueEmployees });
      })
      .on("error", (err) => reject(err));
  });
}

/**
 * Gets unique CPT codes from financial analysis file
 */
export async function getUniqueCPTCodesVitalCare(
  filePath: string
): Promise<string[]> {
  return new Promise((resolve, reject) => {
    const cptCodeSet = new Set<string>();

    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row: any) => {
        const values = Object.values(row) as string[];
        const cptCode = values[2];
        if (cptCode) {
          cptCodeSet.add(cptCode);
        }
      })
      .on("end", () => resolve(Array.from(cptCodeSet)))
      .on("error", (err) => reject(err));
  });
}

/**
 * Processes provider-location relationships for VitalCare
 */
export async function processProviderLocationRelationshipVitalCare(
  filePath: string
): Promise<{
  locationProviderMap: Record<string, string[]>;
  uniqueProviders: string[];
}> {
  return new Promise((resolve, reject) => {
    const locationProviderRelationships: Record<string, string[]> = {};
    const providerSet = new Set<string>();

    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row: any) => {
        const providerName = row["Full Name"]?.trim();
        const locationName = row["Location"]?.trim();

        if (!providerName || !locationName) return;

        if (!locationProviderRelationships[locationName]) {
          locationProviderRelationships[locationName] = [];
        }
        locationProviderRelationships[locationName].push(providerName);
        providerSet.add(providerName);
      })
      .on("end", () => {
        const uniqueProviders = Array.from(providerSet).map((p) => p.trim());
        resolve({
          locationProviderMap: locationProviderRelationships,
          uniqueProviders,
        });
      })
      .on("error", (err) => reject(err));
  });
}

/**
 * Processes accounts receivable CSV file
 */
export async function processAccountsReceivable(
  filePath: string
): Promise<Record<string, any>> {
  return new Promise((resolve, reject) => {
    const result: Record<string, any> = {};
    const columnsToKeep = new Set([
      "0-30 Days",
      "31-60 Days",
      "61-90 Days",
      "91-120 Days",
      "121 - 150 Days",
      "151 - 180 Days",
      "> 180 Days",
      "Total Balance",
      "Total Balance %",
    ]);

    let header: string[] = [];
    let columnIndices: Record<string, number> = {};
    const overallSums: Record<string, number> = {};
    let isFirstRow = true;

    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row: any) => {
        if (isFirstRow) {
          header = Object.keys(row);
          header.forEach((col, idx) => {
            const colName = col.trim();
            if (columnsToKeep.has(colName)) {
              columnIndices[colName] = idx;
            }
          });
          isFirstRow = false;
        }

        const balanceType = Object.values(row)[0] as string;
        if (!balanceType?.trim()) return;

        const balanceData: Record<string, string> = {};
        for (const [colName, idx] of Object.entries(columnIndices)) {
          const value = Object.values(row)[idx] as string;
          if (value?.trim()) {
            balanceData[colName] = value.trim();

            // Parse and sum for Overall row (skip percentages)
            if (colName !== "Total Balance %") {
              const cleanValue = value.replace(/,/g, "");
              const floatVal = parseFloat(cleanValue);
              if (!isNaN(floatVal)) {
                overallSums[colName] = (overallSums[colName] || 0) + floatVal;
              }
            }
          }
        }

        result[balanceType.trim()] = balanceData;
      })
      .on("end", () => {
        // Add "Overall - Sum" row
        const overallData: Record<string, string> = {};
        let totalBalance = 0;

        for (const [colName, sum] of Object.entries(overallSums)) {
          if (colName === "Total Balance") {
            totalBalance = sum;
          }
          overallData[colName] = sum.toFixed(2);
        }

        overallData["Total Balance %"] = "100%";
        result["Overall - Sum"] = overallData;

        // Add "% Subtotal" row
        const subtotalData: Record<string, string> = {};
        if (totalBalance > 0) {
          for (const [colName, sum] of Object.entries(overallSums)) {
            if (colName !== "Total Balance") {
              const percentage = (sum / totalBalance) * 100;
              subtotalData[colName] = `${percentage.toFixed(0)}%`;
            }
          }
        }
        subtotalData["Total Balance"] = "";
        subtotalData["Total Balance %"] = "100%";
        result["% Subtotal"] = subtotalData;

        resolve(result);
      })
      .on("error", (err) => reject(err));
  });
}
